### PR TITLE
[data] Page-lock text batches so the H2D copy is async

### PR DESCRIPTION
**Pull request by SherlockNoMad**
## Summary

`train_step` moved each microbatch to the device with a plain `.to()`. The
batches came out of the collator in pageable memory, and a CUDA copy out of
pageable memory is synchronous with respect to the host: the driver stages it
through its own bounce buffer. Because the copy is also stream-ordered behind
the kernels already queued, the host thread sat inside `.to()` waiting for the
previous gradient-accumulation group's forward and backward to drain, so it
could never run ahead to queue the next one.

`TextCollator` now allocates the padded batch page-locked up front and
concatenates the rows straight into its head, and the trainer copies with
`non_blocking=True`.

Allocating page-locked rather than calling `.pin_memory()` avoids a second
host copy, and allocating with the padding value already in place removes the
separate `F.pad` pass the old code needed. The rows still go in through one
fused `torch.cat(..., out=...)`; filling them in a Python loop costs 2-3x once
a batch holds many rows.

`MultiModalCollator` is left alone. `collate_images` produces the vision
tensors pageable, and the host blocks on the slowest tensor in the batch, so
page-locking only the text half would buy nothing. Multimodal batches behave
exactly as before.

## Measurements

Host time blocked inside `.to()`, GB300, 32 MB batch, ~157 ms of kernels
queued ahead:

| source | flag | host blocked |
| --- | --- | --- |
| pageable | `non_blocking=False` (before) | 157.268 ms |
| pageable | `non_blocking=True` | 157.259 ms |
| page-locked | `non_blocking=True` (after) | 0.093 ms |

The middle row is why the flag alone was not worth landing on its own.

Collation cost against the old `cat`+`pad`, both computing `num_valid_tokens`,
median of 300, 8192-token batch:

| rows | packed (pad=0) | unpacked (pad>0) |
| --- | --- | --- |
| 1 | 1.08x | 0.74x |
| 8 | 1.14x | 0.82x |
| 64 | 1.06x | 0.96x |
| 512 | 1.00x | 1.00x |
| 2048 | 1.01x | 1.00x |

Within a few percent when the batch is exactly full, faster when it is padded,
and flat as the row count grows.

## Behavior change

Tokens are now always int64 rather than inheriting the processor's dtype.
Every in-tree processor already emits int64, it is what the embedding lookup
wants, and a fixed dtype keeps the batch stable for CUDA graph capture. int32
rows still promote rather than erroring.

## Test plan

- `pytest -q tests/unit_tests/cpu --ignore=tests/unit_tests/cpu/test_torch_checkpointing.py`
  -- 808 passed, 1 skipped. The one failure, `test_lora.py::test_lora_forward`,
  is a pre-existing Inductor flex-CPU error that reproduces on a clean tree.
- The same data suites under `CUDA_VISIBLE_DEVICES=""`, since allocating with
  `pin_memory=True` raises without an accelerator -- 124 passed, 1 skipped.
- Collated output is byte-identical to the previous implementation across six
  cases (single row, masked labels, explicit positions, multi-row packed,
  exactly-full batch, random), including dtypes and `num_valid_tokens`.
- `pre-commit run --files <changed files>`

### CODE REVIEW

File: torchtitan/components/data/collators.py [hunk 1]
```diff
@@ -43,7 +49,7 @@ def num_rows_per_batch(self) -> int:
 
 
 class TextCollator(Collator):
```
Comment (tianyu-l) [inline torchtitan/components/data/collators.py:51]: > Collation cost against the old cat+pad, both computing num_valid_tokens,
median of 300, 8192-token batch:
rows	packed (pad=0)	unpacked (pad>0)
1	1.08x	0.74x
8	1.14x	0.82x
64	1.06x	0.96x
512	1.00x	1.00x
2048	1.01x	1.00x
Within a few percent when the batch is exactly full, faster when it is padded,
and flat as the row count grows.

Didn't understand the table / description. Could you explain a bit more?

**Comment by tianyu-l** [comment]
@claude please review this PR

**Review by tianyu-l — APPROVED** [review]
nice, how much acceleration e2e?

**Comment by SherlockNoMad** [comment]
> nice, how much acceleration e2e?

Before the fix, the first H2D copy is blocking the future kernels from launching, and it happens that the following kernels are all small kernels for copies and attention metadata creation, and they are outside of cudgraph region, thus cpu bounded.

image

The overall step time save is ~38ms, so < 0.38%.

File: torchtitan/components/data/collators.py (hunk 1 above)
Comment (SherlockNoMad) [reply torchtitan/components/data/collators.py:51]: explained by claude below.
